### ISSUE TITLE
compact: Ensure downsampled chunks are not larger than 120 samples; stream downsampling more 

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#2542]
During bug fixing on https://github.com/thanos-io/thanos/pull/2528 I found that downsampling always encodes whatever is given in the block into huge chunks. This can lead to inefficiency during query time when only a small part of chunk data is needed, but store GW needs to fetch and decode everything. 

See: https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/pkg/compact/downsample/downsample.go#L141

AC:
* Downsampled chunks are not larger than 120 samples
* Chunks are expanded on demand in iterator.

**Comment by yashrsharma44** [comment]
If no one is working on this issue, can I have a try? @bwplotka

**Comment by bwplotka** [comment]
I assume you are not on it right? (: Just making sure, so others can help as well (:

**Comment by yashrsharma44** [comment]
Yeah, apologies for keeping blocked, I am not working on it currently 😅

**Comment by jyizheng** [comment]
I am happy to work on this.

**Comment by GiedriusS** [comment]
Still relevant AFAIK.

**Comment by jyizheng** [comment]
Hi, I still want to work on this. I will get back to you guy soon.

**Comment by pracucci** [comment]
_Is this still valid? If not, which PR fixed it?_

**Comment by pracucci** [comment]
_I think this is still valid_

**Comment by yeya24** [comment]
Hi, two questions I have about this issue.

1. There is already a limit of 140 samples per chunk for downsampled chunks [here](https://github.com/thanos-io/thanos/blob/master/pkg/compact/downsample/downsample.go#L212). I guess the performance should be similar to 120 samples?

2. What's the benefit of `Chunks are expanded on demand in iterator.`? I see that we need to expand all the chunks before the downsampling anyway.  So what is the doing it `on demand`?

Thanks!

**Comment by pracucci** [comment]
> 1. There is already a limit of 140 samples per chunk for downsampled chunks here. I guess the performance should be similar to 120 samples?

Yes, performances should be similar and 140 should be a problem. However, I'm wondering if that 140 value was a "typo" of 120 or intentional 🤔

**Comment by bwplotka** [comment]
1. Does not matter 120 / 140. If there is check I might miss it and all good (: Let's double check.
2.

> What's the benefit of Chunks are expanded on demand in iterator.? I see that we need to expand all the chunks before the downsampling anyway. So what is the doing it on demand?

What I mean is to not expand all of chunks and only then process query, just instead one, process sample, then another and process those and another etc... (streaming)

**Comment by yeya24** [comment]
Still valid

**Comment by yashrsharma44** [comment]
Hey @bwplotka 
Can you elaborate on this - [link](https://github.com/thanos-io/thanos/issues/2542#issuecomment-728954071)

> What I mean is to not expand all of chunks and only then process query, just instead one, process sample, then another and process those and another etc... (streaming)

**Comment by yeya24** [comment]
#4275 fixes this
